src/sync/treiber_stack.rs

```diff
@@ -49,6 +49,20 @@ impl<T> TreiberStack<T> {
         self.try_pop()
     }
 
+    /// Swaps a stack with a new one.  Can be used to take all the
+    /// elements off the stack at once.
+    ///
+    /// Wait-free
+    pub fn swap(&self, new: TreiberStack<T>) -> TreiberStack<T> {
+        let guard = epoch::pin();
+        // new is exclusively owned and so can be written and read
+        // with relaxed atomics
+        let newhead = new.head.swap(None, Relaxed, &guard);
+        let head = self.head.swap_shared(newhead, AcqRel, &guard);
```

Comment:
Do you mean the `new.head.swap`? I think it is guaranteed to happen on the current thread before `self.head.swap_shared` anyway due to the data dependency of the `swap_shared` on `newhead` and since `new` is only accessible on the current thread, that's enough.